src/services/ApiService.ts

```diff
@@ -200,44 +202,27 @@ export class ApiService {
 
     private upgradePreferences = (preferences: any) => {
         // Upgrade to V2 if required
-        if (preferences?.version && preferences.version === 1) {
-            if (preferences.astColor || preferences.astColor === 0) {
-                switch (preferences.astColor) {
-                    case 0:
-                        preferences.astColor = "auto-black";
-                        break;
-                    case 1:
-                        preferences.astColor = "auto-white";
-                        break;
-                    case 2:
-                        preferences.astColor = "auto-red";
-                        break;
-                    case 3:
-                        preferences.astColor = "auto-forest";
-                        break;
-                    case 4:
-                        preferences.astColor = "auto-blue";
-                        break;
-                    case 5:
-                        preferences.astColor = "auto-turquoise";
-                        break;
-                    case 6:
-                        preferences.astColor = "auto-violet";
-                        break;
-                    case 7:
-                        preferences.astColor = "auto-gold";
-                        break;
-                    case 8:
-                        preferences.astColor = "auto-gray";
-                        break;
-                    default:
-                        preferences.astColor = "auto-blue";
-                        break;
-                }
-            }
-            preferences.version = 2;
+        if (preferences["version"] && preferences["version"] === 1) {
+            // Convert preferences[PreferenceKeys.WCS_OVERLAY_AST_COLOR] from a number in version 1 to a string in version 2
+            // default to "auto-blue" if the value is not in the AST_COLORS map
+            const astColorKey = PreferenceKeys.WCS_OVERLAY_AST_COLOR;
+            if (preferences[astColorKey] || preferences[astColorKey] === 0) {
+                const astColorArrays = Array.from(AST_COLORS.keys());
+                preferences[astColorKey] = astColorArrays.includes(preferences[astColorKey]) ? AST_COLORS.get(preferences[astColorKey]) : "auto-blue";
+            }
+            preferences["version"] = 2;
             this.setPreferences(preferences);
         }
+
+        // Upgrade to schema version 2
+        const schemaVersionRegex = /schema_(\d+)\.json/;
+        const schemaKey = "$schema";
+        if (preferences[schemaKey] && preferences[schemaKey].match(schemaVersionRegex)[1] === "1") {
+            if (preferences[PreferenceKeys.WCS_OVERLAY_WCS_TYPE]) {
+                preferences[PreferenceKeys.WCS_OVERLAY_WCS_TYPE] = preferences[PreferenceKeys.WCS_OVERLAY_WCS_TYPE].toLowerCase();
+                this.setPreferences(preferences);
```

Comment:
As I mentioned in the issue, this (and the other conditional `setPreferences` in this function) will always fail if there are other invalid preferences.

I think that the cleanest way to fix this is to make these requests with the changed value *only* (rather than the entire preferences object). There's already a function for this, `setPreference`, which can be used in both places.

(Both the backend and the controller add a schema version before validation, which is how the function works despite not explicitly including one. We should probably revise that at some point, but it's not necessary right now.)